weave/trace_server/base64_content_conversion.py

```diff
@@ -0,0 +1,509 @@
+                    )
+                    return val
+                else:
+                    return content_obj
+            elif len(val) > MIN_TEXT_SIZE:
+                original_schema = "{text}"
+                return create_content_object(
+                    val.encode('utf-8'),
+                    original_schema,
+                    project_id,
+                    trace_server,
+                    None,
+                    wb_user_id
+                )
+            return val
+        else:
+            return val
+
+    return _visit(vals)
+
+
+R = TypeVar("R", bound=Union[CallStartReq, CallEndReq])
+
+
+def process_call_req_to_content(
+    req: R,
+    trace_server: TraceServerInterface,
+) -> R:
+    """Process call inputs/outputs to replace base64 content.
+
+    This is the main entry point for processing trace data before insertion.
+
+    Args:
+        data: Input or output data from a call
+        project_id: Project ID for storage
+        trace_server: Trace server instance
+
+    Returns:
+        Tuple of (processed_data, list_of_refs) with base64 content replaced by Content objects
+    """
+    if isinstance(req, CallStartReq):
+        req.start.inputs = replace_base64_with_content_objects(
+            req.start.inputs, req.start.project_id, trace_server
+        )
+    else:
+        req.end.output = req.end.output = replace_base64_with_content_objects(
+            req.end.output, req.end.project_id, trace_server
+        )
+
+    return req
+
+
+def reconstruct_base64_for_call(
+    call: CallSchema,
+    trace_server: TraceServerInterface,
+) -> CallSchema:
+    """Recursively reconstruct base64 content from content references.
+
+    Only reconstructs content that has _original_schema metadata, indicating
+    it was originally base64 that we converted. Native Content objects are not
+    reconstructed.
+
+    Args:
+        vals: Value to process (can be dict, list, or primitive)
+        project_id: Project ID for storage
+        trace_server: Trace server instance for file retrieval
+
+    Returns:
+        Processed value with content references replaced by base64 where applicable
+    """
+    import json
+
+    def _visit(val: Any) -> Any:
+        if isinstance(val, dict):
+            # Check if this is a Content object structure
+            if (
+                val.get("_type") == "CustomWeaveType"
+                and val.get("weave_type", {}).get("type")
+                == "weave.type_wrappers.Content.content.Content"
+                and "files" in val
+            ):
+                files = val["files"]
+
+                # We expect both content and metadata.json digests
+                if "content" not in files or "metadata.json" not in files:
+                    # Not a complete content object, process normally
+                    result = {}
+                    for k, v in val.items():
+                        result[k] = _visit(v)
```

Comment:
Nit
```suggestion
                        result = {
                            k: _visit(v) for k, v in val.items()
                        }
```

Or even more simply:
```python
                        result = _visit(val)
```
since `_visit` walks dicts.